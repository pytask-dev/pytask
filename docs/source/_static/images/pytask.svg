<?xml version="1.0" encoding="utf-8"?>
<!-- Generator: Adobe Illustrator 24.1.1, SVG Export Plug-In . SVG Version: 6.00 Build 0)  -->
<svg version="1.1" id="Capa_1" xmlns="http://www.w3.org/2000/svg" xmlns:xlink="http://www.w3.org/1999/xlink" x="0px" y="0px"
	 viewBox="0 0 438.9 438.9" style="enable-background:new 0 0 438.9 438.9;" xml:space="preserve">
<style type="text/css">
	.st0{fill:#C19C00;}
	.st1{fill:#13A10E;}
	.st2{fill:#C50F1F;}
</style>
<path d="M348,57.5h-39.7V39.7c0-5.7-6.3-8.4-12-8.4h-30.8C258.1,10.5,239.8,0,218.9,0c-20.7-0.8-39.5,11.9-46.5,31.3h-30.3
	c-5.7,0-11.5,2.6-11.5,8.4v17.8H90.9C67.4,57.8,48.1,76.3,47,99.8v299.4c0,23,20.9,39.7,43.9,39.7h257c23,0,43.9-16.7,43.9-39.7
	V99.8C390.7,76.3,371.5,57.8,348,57.5z M151.5,52.3h28.7c5-0.6,9-4.4,9.9-9.4c3.1-13.5,14.9-23.1,28.7-23.5
	c13.7,0.4,25.3,10.1,28.2,23.5c0.9,5.1,5.2,9,10.4,9.4h29.8v41.8H151.5V52.3z M371,399.2c0,11.5-11.5,18.8-23,18.8h-257
	c-11.5,0-23-7.3-23-18.8V99.8c1.1-12,11-21.2,23-21.4h39.7V105c0.6,5.9,5.6,10.2,11.5,9.9h154.1c6,0.3,11.2-4,12-9.9V78.4H348
	c12,0.2,21.9,9.5,23,21.4L371,399.2L371,399.2z"/>
<path class="st0" d="M179.2,233.6c-3.9-4.1-10.4-4.4-14.6-0.5l-33.4,31.9L117,250.3c-3.9-4.1-10.4-4.4-14.6-0.5
	c-4,4.2-4,10.9,0,15.2l21.4,21.9c1.9,2.1,4.5,3.2,7.3,3.1c2.8,0,5.4-1.2,7.3-3.1l40.8-38.7c4-3.7,4.3-10,0.6-14
	C179.6,234,179.4,233.8,179.2,233.6z"/>
<path d="M329.2,256H209c-5.8,0-10.4,4.7-10.4,10.4s4.7,10.4,10.4,10.4h120.2c5.8,0,10.4-4.7,10.4-10.4S334.9,256,329.2,256z"/>
<path class="st1" d="M179.2,150c-3.9-4.1-10.4-4.4-14.6-0.5l-33.4,31.9L117,166.7c-3.9-4.1-10.4-4.4-14.6-0.5c-4,4.2-4,10.9,0,15.2
	l21.4,21.9c1.9,2.1,4.5,3.2,7.3,3.1c2.8,0,5.4-1.2,7.3-3.1l40.8-38.7c4-3.7,4.3-10,0.6-14C179.6,150.4,179.4,150.2,179.2,150z"/>
<path d="M329.2,172.4H209c-5.8,0-10.4,4.7-10.4,10.4s4.7,10.4,10.4,10.4h120.2c5.8,0,10.4-4.7,10.4-10.4S334.9,172.4,329.2,172.4z"
	/>
<path class="st2" d="M179.2,317.2c-3.9-4.1-10.4-4.4-14.6-0.5l-33.4,31.9L117,333.9c-3.9-4.1-10.4-4.4-14.6-0.5
	c-4,4.2-4,10.9,0,15.2l21.4,21.9c1.9,2.1,4.5,3.2,7.3,3.1c2.8,0,5.4-1.2,7.3-3.1l40.8-38.7c4-3.7,4.3-10,0.6-14
	C179.6,317.6,179.4,317.4,179.2,317.2z"/>
<path d="M329.2,339.6H209c-5.8,0-10.4,4.7-10.4,10.4s4.7,10.4,10.4,10.4h120.2c5.8,0,10.4-4.7,10.4-10.4S334.9,339.6,329.2,339.6z"
	/>
</svg>
