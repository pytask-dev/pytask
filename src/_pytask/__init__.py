"""This module should not contain any imports except for the version."""
from __future__ import annotations


try:
    from ._version import version as __version__
except ImportError:  # pragma: no cover
    # broken installation, we don't even try unknown only works because we do poor mans
    # version compare
    __version__ = "unknown"


__all__ = ["__version__"]
