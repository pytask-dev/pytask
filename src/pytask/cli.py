import sys
from pathlib import Path

import click
from pytask.config import hookimpl
from pytask.pluginmanager import get_plugin_manager


CONTEXT_SETTINGS = {"help_option_names": ["-h", "--help"]}


def add_parameters(func):
    """Add parameters from plugins to the commandline interface."""
    pm = _prepare_plugin_manager()
    pm.hook.pytask_add_parameters_to_cli(command=func)
    # Hack to pass the plugin manager via a hidden option to the ``config_from_cli``.
    func.params.append(click.Option(["--pm"], default=pm, hidden=True))

    return func


def _prepare_plugin_manager():
    pm = get_plugin_manager()
    pm.register(sys.modules[__name__])
    pm.hook.pytask_add_hooks(pm=pm)
    return pm


@hookimpl
def pytask_add_hooks(pm):
    from pytask import collect
    from pytask import config
    from pytask import database
    from pytask import debugging
    from pytask import execute
    from pytask import logging
    from pytask import main
    from pytask import parametrize
    from pytask import resolve_dependencies
    from pytask import skipping
    from pytask import mark_

    pm.register(collect)
    pm.register(config)
    pm.register(database)
    pm.register(debugging)
    pm.register(execute)
    pm.register(logging)
    pm.register(main)
    pm.register(parametrize)
    pm.register(resolve_dependencies)
    pm.register(skipping)
    pm.register(mark_)


def _to_path(ctx, param, value):  # noqa: U100
    """Callback for :class:`click.Argument` or :class:`click.Option`."""
    return [Path(i).resolve() for i in value]


@hookimpl
def pytask_add_parameters_to_cli(command):
    additional_parameters = [
        click.Argument(
            ["paths"], nargs=-1, type=click.Path(exists=True), callback=_to_path,
        ),
        click.Option(
            ["--ignore"],
            type=str,
            multiple=True,
            help="Ignore path (globs and multi allowed).",
        ),
        click.Option(
            ["--debug-pytask"],
            is_flag=True,
            help="Debug pytask by tracing all hook calls.",
        ),
    ]
    command.params.extend(additional_parameters)


@add_parameters
@click.command(context_settings=CONTEXT_SETTINGS)
@click.version_option()
def pytask(**config_from_cli):
    """Command-line interface for pytask."""
    session = main(config_from_cli)
    sys.exit(session.exit_code)


def main(config_from_cli):
    pm = config_from_cli.get("pm", _prepare_plugin_manager())
    session = pm.hook.pytask_main(config_from_cli=config_from_cli)

    return session
