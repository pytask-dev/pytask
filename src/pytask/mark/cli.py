import textwrap

import click
import pytask
from pytask import main
from pytask.enums import ExitCode
from pytask.main import Session
from pytask.pluginmanager import get_plugin_manager


@pytask.hookimpl
def pytask_add_parameters_to_cli(command: click.Command) -> None:
    command.params.append(
        click.Option(["--markers"], is_flag=True, help="Show available markers.")
    )


@pytask.hookimpl(tryfirst=True)
def pytask_main(config_from_cli: dict) -> int:
    if config_from_cli.get("markers", False):
        try:
            # Duplication of the same mechanism in :func:`pytask.main.pytask_main`.
            pm = get_plugin_manager()
            pm.register(main)
            pm.hook.pytask_add_hooks(pm=pm)

            config = pm.hook.pytask_configure(pm=pm, config_from_cli=config_from_cli)

            session = Session.from_config(config)
            session.exit_code = ExitCode.OK

        except Exception as e:
            raise Exception("Error while configuring pytask.") from e

        for name, description in config["markers"].items():
            click.echo(
                textwrap.fill(
                    f"pytask.mark.{name}: {description}", width=config["terminal_width"]
                )
            )
            click.echo("")

        return session
