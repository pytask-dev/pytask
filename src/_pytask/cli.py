"""Implements the command line interface."""
from __future__ import annotations

import sys
from typing import Any

import click
import pluggy
from _pytask.click import ColoredGroup
from _pytask.config import hookimpl
from _pytask.pluginmanager import get_plugin_manager
from packaging.version import parse as parse_version


_CONTEXT_SETTINGS: dict[str, Any] = {"help_option_names": ("-h", "--help")}


if parse_version(click.__version__) < parse_version("8"):
    _VERSION_OPTION_KWARGS: dict[str, Any] = {}
else:
    _VERSION_OPTION_KWARGS = {"package_name": "pytask"}


def _extend_command_line_interface(cli: click.Group) -> click.Group:
    """Add parameters from plugins to the commandline interface."""
    pm = _prepare_plugin_manager()
    pm.hook.pytask_extend_command_line_interface(cli=cli)
    _sort_options_for_each_command_alphabetically(cli)
    return cli


def _prepare_plugin_manager() -> pluggy.PluginManager:
    """Prepare the plugin manager."""
    pm = get_plugin_manager()
    pm.register(sys.modules[__name__])
    pm.hook.pytask_add_hooks(pm=pm)
    return pm


def _sort_options_for_each_command_alphabetically(cli: click.Group) -> None:
    """Sort command line options and arguments for each command alphabetically."""
    for command in cli.commands:
        cli.commands[command].params = sorted(
            cli.commands[command].params, key=lambda x: x.name
        )


@hookimpl
def pytask_add_hooks(pm: pluggy.PluginManager) -> None:
    """Add hooks."""
    from _pytask import build
    from _pytask import capture
    from _pytask import clean
    from _pytask import collect
    from _pytask import collect_command
    from _pytask import config
    from _pytask import database
    from _pytask import debugging
    from _pytask import execute
    from _pytask import graph
    from _pytask import live
    from _pytask import logging
    from _pytask import mark
    from _pytask import parameters
    from _pytask import parametrize
    from _pytask import persist
    from _pytask import profile
    from _pytask import resolve_dependencies
    from _pytask import skipping
    from _pytask import task
    from _pytask import warnings

    pm.register(build)
    pm.register(capture)
    pm.register(clean)
    pm.register(collect)
    pm.register(collect_command)
    pm.register(config)
    pm.register(database)
    pm.register(debugging)
    pm.register(execute)
    pm.register(graph)
    pm.register(live)
    pm.register(logging)
    pm.register(mark)
    pm.register(parameters)
    pm.register(parametrize)
    pm.register(persist)
    pm.register(profile)
    pm.register(resolve_dependencies)
    pm.register(skipping)
    pm.register(task)
    pm.register(warnings)


@click.group(
    cls=ColoredGroup,
    context_settings=_CONTEXT_SETTINGS,
    default="build",
    default_if_no_args=True,
)
@click.version_option(**_VERSION_OPTION_KWARGS)
def cli() -> None:
    """Manage your tasks with pytask."""
    pass


_extend_command_line_interface(cli)
