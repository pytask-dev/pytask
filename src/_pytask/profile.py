"""This module contains the code to profile the execution."""
from __future__ import annotations

import csv
import json
import sys
import time
from enum import Enum
from pathlib import Path
from types import TracebackType
from typing import Any
from typing import Generator
from typing import TYPE_CHECKING

import click
from _pytask.click import ColoredCommand
from _pytask.config import hookimpl
from _pytask.config_utils import parse_click_choice
from _pytask.console import console
from _pytask.console import format_task_id
from _pytask.database import db
from _pytask.exceptions import CollectionError
from _pytask.exceptions import ConfigurationError
from _pytask.nodes import FilePathNode
from _pytask.nodes import Task
from _pytask.outcomes import ExitCode
from _pytask.outcomes import TaskOutcome
from _pytask.pluginmanager import get_plugin_manager
from _pytask.report import ExecutionReport
from _pytask.session import Session
from _pytask.shared import get_first_non_none_value
from _pytask.traceback import render_exc_info
from pony import orm
from rich.table import Table


if TYPE_CHECKING:
    from typing import NoReturn


class _ExportFormats(Enum):
    NO = "no"
    JSON = "json"
    CSV = "csv"


class Runtime(db.Entity):  # type: ignore
    """Record of runtimes of tasks."""

    task = orm.PrimaryKey(str)
    date = orm.Required(float)
    duration = orm.Required(float)


@hookimpl(tryfirst=True)
def pytask_extend_command_line_interface(cli: click.Group) -> None:
    """Extend the command line interface."""
    cli.add_command(profile)


@hookimpl
def pytask_parse_config(
    config: dict[str, Any], config_from_cli: dict[str, Any]
) -> None:
    """Parse the configuration."""
    config["export"] = get_first_non_none_value(
        config_from_cli,
        key="export",
        default=_ExportFormats.NO,
        callback=parse_click_choice("export", _ExportFormats),
    )


@hookimpl
def pytask_post_parse(config: dict[str, Any]) -> None:
    """Register the export option."""
    config["pm"].register(ExportNameSpace)
    config["pm"].register(DurationNameSpace)
    config["pm"].register(FileSizeNameSpace)


@hookimpl(hookwrapper=True)
def pytask_execute_task(task: Task) -> Generator[None, None, None]:
    """Attach the duration of the execution to the task."""
    start = time.time()
    yield
    end = time.time()
    task.attributes["duration"] = (start, end)


@hookimpl
def pytask_execute_task_process_report(report: ExecutionReport) -> None:
    """Store runtime of successfully finishing tasks in database."""
    task = report.task
    duration = task.attributes.get("duration")
    if report.outcome == TaskOutcome.SUCCESS and duration is not None:
        _create_or_update_runtime(task.name, *duration)


@orm.db_session
def _create_or_update_runtime(task_name: str, start: float, end: float) -> None:
    """Create or update a runtime entry."""
    try:
        runtime = Runtime[task_name]  # type: ignore
    except orm.ObjectNotFound:
        Runtime(task=task_name, date=start, duration=end - start)
    else:
        for attr, val in [("date", start), ("duration", end - start)]:
            setattr(runtime, attr, val)


@click.command(cls=ColoredCommand)
@click.option(
    "--export",
    type=click.Choice([x.value for x in _ExportFormats]),
    default=None,
    help="Export the profile in the specified format. [dim]\\[default: no][/]",
)
def profile(**config_from_cli: Any) -> NoReturn:
    """Show information about tasks like runtime and memory consumption of products."""
    config_from_cli["command"] = "profile"

    try:
        # Duplication of the same mechanism in :func:`pytask.main.main`.
        pm = get_plugin_manager()
        from _pytask import cli

        pm.register(cli)
        pm.hook.pytask_add_hooks(pm=pm)

        config = pm.hook.pytask_configure(pm=pm, config_from_cli=config_from_cli)
        session = Session.from_config(config)

    except (ConfigurationError, Exception):  # pragma: no cover
        session = Session({}, None)
        session.exit_code = ExitCode.CONFIGURATION_FAILED
        exc_info: tuple[
            type[BaseException], BaseException, TracebackType | None
        ] = sys.exc_info()
        console.print(render_exc_info(*exc_info, show_locals=config["show_locals"]))

    else:
        try:
            session.hook.pytask_log_session_header(session=session)
            session.hook.pytask_collect(session=session)
            session.hook.pytask_resolve_dependencies(session=session)

            profile: dict[str, dict[str, Any]] = {
                task.name: {} for task in session.tasks
            }
            session.hook.pytask_profile_add_info_on_task(
                session=session, tasks=session.tasks, profile=profile
            )
            profile = _process_profile(profile)

            _print_profile_table(profile, session.tasks, session.config)

            session.hook.pytask_profile_export_profile(session=session, profile=profile)

            console.rule(style="neutral")

        except CollectionError:  # pragma: no cover
            session.exit_code = ExitCode.COLLECTION_FAILED

        except Exception:  # pragma: no cover
            session.exit_code = ExitCode.FAILED
            console.print_exception()
            console.rule(style="failed")

    sys.exit(session.exit_code)


def _print_profile_table(
    profile: dict[str, dict[str, Any]], tasks: list[Task], config: dict[str, Any]
) -> None:
    """Print the profile table."""
    name_to_task = {task.name: task for task in tasks}
    info_names = _get_info_names(profile)

    console.print()
    if profile:
        table = Table("Task")
        for name in info_names:
            table.add_column(name, justify="right")

        for task_name, info in profile.items():
            task_id = format_task_id(
                task=name_to_task[task_name],
                editor_url_scheme=config["editor_url_scheme"],
                short_name=True,
            )
            infos = [str(i) for i in info.values()]
            table.add_row(task_id, *infos)

        console.print(table)
    else:
        console.print("No information is stored on the collected tasks.")


class DurationNameSpace:
    """A namespace for adding durations to the profile."""

    @staticmethod
    @hookimpl
    def pytask_profile_add_info_on_task(
        tasks: list[Task], profile: dict[str, dict[str, Any]]
    ) -> None:
        """Add the runtime for tasks to the profile."""
        runtimes = _collect_runtimes([task.name for task in tasks])
        for name, duration in runtimes.items():
            profile[name]["Duration (in s)"] = round(duration, 2)


@orm.db_session
def _collect_runtimes(task_names: list[str]) -> dict[str, float]:
    """Collect runtimes."""
    runtimes = [Runtime.get(task=task_name) for task_name in task_names]
    runtimes = [r for r in runtimes if r is not None]
    return {r.task: r.duration for r in runtimes}


class FileSizeNameSpace:
    """A namespace for adding the total file size of products to a task."""

    @staticmethod
    @hookimpl
    def pytask_profile_add_info_on_task(
        session: Session, tasks: list[Task], profile: dict[str, dict[str, Any]]
    ) -> None:
        """Add the total file size of all products for a task."""
        for task in tasks:
            successors = list(session.dag.successors(task.name))
            if successors:
                sum_bytes = 0
                for successor in successors:
                    node = session.dag.nodes[successor]["node"]
                    if isinstance(node, FilePathNode):
                        try:
                            sum_bytes += node.path.stat().st_size
                        except FileNotFoundError:
                            pass

                profile[task.name]["Size of Products"] = _to_human_readable_size(
                    sum_bytes
                )


def _to_human_readable_size(bytes_: int, units: list[str] | None = None) -> str:
    """Convert bytes to a human readable size."""
    units = [" bytes", " KB", " MB", " GB", " TB"] if units is None else units
    return (
        str(bytes_) + units[0]
        if bytes_ < 1024 or len(units) == 1
        else _to_human_readable_size(bytes_ >> 10, units[1:])
    )


def _process_profile(profile: dict[str, dict[str, Any]]) -> dict[str, dict[str, Any]]:
    """Process profile to make it ready for printing and storing."""
    info_names = _get_info_names(profile)
    if info_names:
        complete_profiles = {
            task_name: {
                attr_name: profile[task_name].get(attr_name, "")
                for attr_name in info_names
            }
            for task_name in sorted(profile)
        }
    else:
        complete_profiles = {}
    return complete_profiles


class ExportNameSpace:
    """A namespace for exporting the profile."""

    @staticmethod
    @hookimpl(trylast=True)
    def pytask_profile_export_profile(
        session: Session, profile: dict[str, dict[str, Any]]
    ) -> None:
        """Export profiles."""
        export = session.config["export"]

        if export == _ExportFormats.CSV:
            _export_to_csv(profile)
        elif export == _ExportFormats.JSON:
            _export_to_json(profile)
        elif export == _ExportFormats.NO:
            pass
        else:  # pragma: no cover
            raise ValueError(f"The export option {export.value!r} cannot be handled.")


def _export_to_csv(profile: dict[str, dict[str, Any]]) -> None:
    """Export profile to csv."""
    info_names = _get_info_names(profile)
    path = Path.cwd().joinpath("profile.csv")

    with open(path, "w", newline="") as file:
        writer = csv.writer(file)
        writer.writerow(("Task", *info_names))
        for task_name, info in profile.items():
            writer.writerow((task_name, *info.values()))


def _export_to_json(profile: dict[str, dict[str, Any]]) -> None:
    """Export profile to json."""
    json_ = json.dumps(profile)
    path = Path.cwd().joinpath("profile.json")
    path.write_text(json_)


def _get_info_names(profile: dict[str, dict[str, Any]]) -> list[str]:
    """Get names of infos of tasks.

    Examples
    --------
    >>> _get_info_names({"t1": {"time": 1}, "t2": {"time": 1, "size": "2GB"}})
    ['size', 'time']
    >>> _get_info_names({})
    []
    >>> _get_info_names({"t1": {}})
    []

    """
    base: set[str] = set()
    info_names: list[str] = sorted(base.union(*(set(val) for val in profile.values())))
    return info_names
