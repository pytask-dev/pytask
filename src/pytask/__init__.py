"""This module contains the main namespace for pytask."""
from __future__ import annotations

from _pytask import __version__
from _pytask.build import main
from _pytask.collect_utils import depends_on
from _pytask.collect_utils import parse_nodes
from _pytask.collect_utils import produces
from _pytask.compat import check_for_optional_program
from _pytask.compat import import_optional_dependency
from _pytask.config import hookimpl
from _pytask.console import console
from _pytask.database import db
from _pytask.exceptions import CollectionError
from _pytask.exceptions import ConfigurationError
from _pytask.exceptions import ExecutionError
from _pytask.exceptions import NodeNotCollectedError
from _pytask.exceptions import NodeNotFoundError
from _pytask.exceptions import PytaskError
from _pytask.exceptions import ResolvingDependenciesError
from _pytask.graph import build_dag
from _pytask.mark import Mark
from _pytask.mark import MARK_GEN as mark  # noqa: N811
from _pytask.mark import MarkDecorator
from _pytask.mark import MarkGenerator
from _pytask.mark_utils import get_all_marks
from _pytask.mark_utils import get_marks
from _pytask.mark_utils import has_mark
from _pytask.mark_utils import remove_marks
from _pytask.mark_utils import set_marks
from _pytask.models import CollectionMetadata
from _pytask.nodes import FilePathNode
from _pytask.nodes import MetaNode
from _pytask.nodes import Task
from _pytask.outcomes import CollectionOutcome
from _pytask.outcomes import count_outcomes
from _pytask.outcomes import Exit
from _pytask.outcomes import ExitCode
from _pytask.outcomes import Persisted
from _pytask.outcomes import Skipped
from _pytask.outcomes import SkippedAncestorFailed
from _pytask.outcomes import SkippedUnchanged
from _pytask.outcomes import TaskOutcome
from _pytask.report import CollectionReport
from _pytask.report import ExecutionReport
from _pytask.report import ResolvingDependenciesReport
from _pytask.session import Session
from _pytask.traceback import format_exception_without_traceback
from _pytask.traceback import remove_internal_traceback_frames_from_exc_info
from _pytask.traceback import remove_traceback_from_exc_info
from _pytask.traceback import render_exc_info
from _pytask.warnings_utils import parse_warning_filter
from _pytask.warnings_utils import warning_record_to_str
from _pytask.warnings_utils import WarningReport


# This import must come last, otherwise a circular import occurs.
from _pytask.cli import cli  # noreorder


__all__ = [
    "__version__",
    "build_dag",
    "check_for_optional_program",
    "cli",
    "console",
    "count_outcomes",
    "db",
    "depends_on",
    "format_exception_without_traceback",
    "get_all_marks",
    "get_first_non_none_value",
    "get_marks",
    "has_mark",
    "hookimpl",
    "import_optional_dependency",
    "main",
    "mark",
    "parse_nodes",
    "parse_warning_filter",
    "produces",
    "remove_internal_traceback_frames_from_exc_info",
    "remove_marks",
    "remove_traceback_from_exc_info",
    "render_exc_info",
    "set_marks",
    "warning_record_to_str",
    "CollectionError",
    "CollectionMetadata",
    "CollectionOutcome",
    "CollectionReport",
    "ConfigurationError",
    "ExecutionReport",
    "ExecutionError",
    "Exit",
    "ExitCode",
    "FilePathNode",
    "Mark",
    "MarkDecorator",
    "MarkGenerator",
    "MetaNode",
    "NodeNotCollectedError",
    "NodeNotFoundError",
    "Persisted",
    "PytaskError",
    "ResolvingDependenciesError",
    "ResolvingDependenciesReport",
    "Session",
    "Skipped",
    "SkippedAncestorFailed",
    "SkippedUnchanged",
    "Task",
    "TaskOutcome",
    "WarningReport",
]
