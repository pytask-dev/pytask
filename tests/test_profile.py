from __future__ import annotations

import os
import textwrap
from pathlib import Path

import pytest
from _pytask.cli import cli
from _pytask.profile import _to_human_readable_size
from _pytask.profile import Runtime
from pytask import create_database
from pytask import DatabaseSession
from pytask import ExitCode
from pytask import main


@pytest.mark.end_to_end()
def test_duration_is_stored_in_task(tmp_path):
    source = """
    import time
    def task_example(): time.sleep(2)
    """
    tmp_path.joinpath("task_example.py").write_text(textwrap.dedent(source))

    session = main({"paths": tmp_path})

    assert session.exit_code == ExitCode.OK
    assert len(session.tasks) == 1
    task = session.tasks[0]
    duration = task.attributes["duration"]
    assert duration[1] - duration[0] > 2

    create_database("sqlite:///" + tmp_path.joinpath(".pytask.sqlite3").as_posix())

    with DatabaseSession() as session:
        task_name = tmp_path.joinpath("task_example.py").as_posix() + "::task_example"

        runtime = session.get(Runtime, task_name)
        assert runtime.duration > 2


@pytest.mark.end_to_end()
def test_profile_if_no_tasks_are_collected(tmp_path, runner):
    result = runner.invoke(cli, ["profile", tmp_path.as_posix()])

    assert result.exit_code == ExitCode.OK
    assert "No information is stored on the collected tasks." in result.output


@pytest.mark.end_to_end()
def test_profile_if_there_is_no_information_on_collected_tasks(tmp_path, runner):
    source = """
    import time
    def task_example(): time.sleep(2)
    """
    tmp_path.joinpath("task_example.py").write_text(textwrap.dedent(source))

    result = runner.invoke(cli, ["profile", tmp_path.as_posix()])

    assert result.exit_code == ExitCode.OK
    assert "Collected 1 task." in result.output
    assert "No information is stored on the collected tasks." in result.output


@pytest.mark.end_to_end()
def test_profile_if_there_is_information_on_collected_tasks(tmp_path, runner):
    source = """
    import time
    import pytask

    @pytask.mark.produces("out.txt")
    def task_example(produces):
        time.sleep(2)
        produces.write_text("There are nine billion bicycles in Beijing.")
    """
    tmp_path.joinpath("task_example.py").write_text(textwrap.dedent(source))

    result = runner.invoke(cli, [tmp_path.as_posix()])

    result = runner.invoke(cli, ["profile", tmp_path.as_posix()])

    assert result.exit_code == ExitCode.OK
    assert "Collected 1 task." in result.output
    assert "Duration (in s)" in result.output
    assert "0." in result.output
    assert "Size of Products" in result.output
    assert "43 bytes" in result.output


@pytest.mark.end_to_end()
@pytest.mark.parametrize("export", ["csv", "json"])
def test_export_of_profile(tmp_path, runner, export):
    source = """
    import time
    def task_example(): time.sleep(2)
    """
    tmp_path.joinpath("task_example.py").write_text(textwrap.dedent(source))

    result = runner.invoke(cli, [tmp_path.as_posix()])

    cwd = Path.cwd()
    os.chdir(tmp_path)
    result = runner.invoke(cli, ["profile", tmp_path.as_posix(), "--export", export])
    os.chdir(cwd)

    assert result.exit_code == ExitCode.OK
    assert "Collected 1 task." in result.output
    assert "Duration (in s)" in result.output
    assert "0." in result.output
    assert tmp_path.joinpath(f"profile.{export}").exists()


@pytest.mark.unit()
@pytest.mark.parametrize(
    ("bytes_", "units", "expected"),
    [
        (2**10, None, "1 KB"),
        (2**20, None, "1 MB"),
        (2**30, None, "1 GB"),
        (2**30, [" bytes", " KB", " MB"], "1024 MB"),
    ],
)
def test_to_human_readable_size(bytes_, units, expected):
    result = _to_human_readable_size(bytes_, units)
    assert result == expected
