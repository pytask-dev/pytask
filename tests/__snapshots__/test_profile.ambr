# serializer version: 1
# name: test_export_of_profile[csv]
  '''
  ───────────────────────────── Start pytask session ─────────────────────────────
  Platform: <platform> -- Python <version>, pytask <version>, pluggy <version>
  Root: <path>
  Collected 1 task.
  
  ┏━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━┳━━━━━━━━━━━━━━━━━┓
  ┃ Task                          ┃ Duration (in s) ┃
  ┡━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━╇━━━━━━━━━━━━━━━━━┩
  │ task_example.py::task_example │             2.0 │
  └───────────────────────────────┴─────────────────┘
  ────────────────────────────────────────────────────────────────────────────────
  '''
# ---
# name: test_export_of_profile[json]
  '''
  ───────────────────────────── Start pytask session ─────────────────────────────
  Platform: <platform> -- Python <version>, pytask <version>, pluggy <version>
  Root: <path>
  Collected 1 task.
  
  ┏━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━┳━━━━━━━━━━━━━━━━━┓
  ┃ Task                          ┃ Duration (in s) ┃
  ┡━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━╇━━━━━━━━━━━━━━━━━┩
  │ task_example.py::task_example │             2.0 │
  └───────────────────────────────┴─────────────────┘
  ────────────────────────────────────────────────────────────────────────────────
  '''
# ---
# name: test_profile_if_there_is_information_on_collected_tasks
  '''
  ───────────────────────────── Start pytask session ─────────────────────────────
  Platform: <platform> -- Python <version>, pytask <version>, pluggy <version>
  Root: <path>
  Collected 1 task.
  
  ┏━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━┳━━━━━━━━━━━━━━━━━┳━━━━━━━━━━━━━━━━━━┓
  ┃ Task                          ┃ Duration (in s) ┃ Size of Products ┃
  ┡━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━╇━━━━━━━━━━━━━━━━━╇━━━━━━━━━━━━━━━━━━┩
  │ task_example.py::task_example │             2.0 │         43 bytes │
  └───────────────────────────────┴─────────────────┴──────────────────┘
  ────────────────────────────────────────────────────────────────────────────────
  '''
# ---
